### PR TITLE
st2 [ADD][1847] purchase_report_custom_hls: 

### PR DESCRIPTION
**Pull request by TomohiroAbe-Quartile**
[1847](https://www.quartile.co/web#id=1847&action=771&model=project.task&view_type=form&menu_id=505)
In Japanese: Company Representative
japanesePO

In English: Company Representative
englishPO

Add payment term field.
スクリーンショット 2021-09-15 125435

### CODE REVIEW

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
I want to replace this line.

purchase/report/purchase_order_templates.xml:3-20

```
<template id="report_purchaseorder_document">
    <t t-call="web.external_layout">
        <t t-set="o" t-value="o.with_context(lang=o.partner_id.lang)"/>
        <t t-set="address">
            <div t-field="o.partner_id"
            t-options='{"widget": "contact", "fields": ["address", "name", "phone"], "no_marker": True, "phone_icons": True}'/>
            <p t-if="o.partner_id.vat"><t t-esc="o.company_id.country_id.vat_label or 'Tax ID'"/>: <span t-field="o.partner_id.vat"/></p>
        </t>
        <t t-if="o.dest_address_id">
            <t t-set="information_block">
                <strong>Shipping address:</strong>
                <div t-if="o.dest_address_id">
                    <div t-field="o.dest_address_id"
                        t-options='{"widget": "contact", "fields": ["address", "name", "phone"], "no_marker": True, "phone_icons": True}' name="purchase_shipping_address"/>
                </div>
            </t>
        </t>
```

But I couldn't replace to use xpath.
`<xpath expr="//div[@t-field='o.partner_id']" position="replace">`

If you have any solution, please help me.

File: purchase_report_custom_hls/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+from odoo import fields, models
+
+
+class Partner(models.Model):
+    _inherit = "res.partner"
+
+    customer_incoterm_place = fields.Char(
+        "Default Incoterm Place",
+        help="The default incoterm place for purchase report.",
+    )
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/res_partner.py]: Is there a requirement to keep the default incoterm place in partner?  Why customer?

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            <div t-if="o.payment_term_id">
+                <strong>Payment Term:</strong>
+                <br />
+                <span t-field="o.payment_term_id.name"/>
+            </div>
+        </xpath>
+        <xpath expr="//div[@t-field='o.partner_id']" position="replace">
+            <div t-field="o.partner_id"
+            t-options='{"widget": "contact", "fields": ["purchase_address", "name", "phone"], "no_marker": True, "phone_icons": True}'/>
+        </xpath>
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: What are you trying to achieve with this? Maybe you are confused between the company address and the partner (supplier) address?

Comment (yostashiro) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I was thinking that you could simply use this module https://github.dev/qrtl/rpl-custom/blob/e56c56f8d98bb66c6209cc374719c1c8f2c43b38/report_company_address to show the company address in English, depending on the partner's language.

Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Maybe I'd misunderstanding. Thank you for your help.

File: purchase_report_custom_hls/models/res_partner.py (hunk 1 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/res_partner.py]: Maybe not use, Thank you.

File: purchase_report_custom_hls/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Purchase Order Report Custom",
+    "version": "12.0.1.0.0",
+    "category": "Purchase",
+    "author": "Quartile Limited",
+    "website": "https://www.quartile.co",
+    "license": "LGPL-3",
+    "installable": True,
+    "depends": ["purchase", "web"],
```
Comment (yostashiro) [inline purchase_report_custom_hls/__manifest__.py]:
```suggestion
    "depends": ["purchase"],
```
`web` is a distant dependency of `purchase`, therefore no need to be included.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                                alt="Signature"
+                            />
+                        </div>
+                        <hr
+                            width="1000%"
+                            style="border: none; border-bottom: 2px solid black;"
+                        />
+                    </td>
+                </tr>
+
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]:
```suggestion
```

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" ?>
+<odoo>
+    <template
+        id="report_purchaseorder_document_inherit_purchase_order"
+        inherit_id="purchase.report_purchaseorder_document"
+    >
+        <xpath expr="//div[@t-if='o.date_order']" position="before">
+            <div t-if="o.payment_term_id">
+                <strong>Payment Term:</strong>
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]:
```suggestion
                <strong>Payment Terms:</strong>
```
Just to make it consistent with the sales side.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                <br />
+                <span t-field="o.payment_term_id.name"/>
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]:
```suggestion
                <span t-field="o.payment_term_id"/>
```

File: purchase_report_custom_hls/models/res_company.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2021 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    company_representative = fields.Char("Company Representative", translate=True,)
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/res_company.py]:
```suggestion
    rep_title = fields.Text("Title of the Representative", translate=True, help="The input value is may show as the title of the representative in some of the printed documents.")
```

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+                                t-if="o.company_id.rep_signature"
+                                t-att-src="image_data_uri(o.company_id.rep_signature)"
+                                style="max-height: 40px;"
+                                alt="Signature"
+                            />
+                        </div>
+                        <hr
+                            width="1000%"
+                            style="border: none; border-bottom: 2px solid black;"
+                        />
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Show `rep_title` below the border line.

File: report_common_hls/models/res_company.py [hunk 9]
```diff
@@ -9,3 +9,6 @@ class ResCompany(models.Model):
 
     company_chop = fields.Binary("Company Chop Image", attachment=True,)
     fax = fields.Char("Fax")
+    purchase_address = fields.Char(
+        "Purchase Address", help="Address for print purchase report."
+    )
```
Comment (yostashiro) [inline report_common_hls/models/res_company.py]: What is this field for?

Comment (TomohiroAbe-Quartile) [reply report_common_hls/models/res_company.py]: I delete this, sorry.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 8 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: スクリーンショット 2021-09-15 184252

Here the current layout.

File: purchase_report_custom_hls/models/res_company.py (hunk 7 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/res_company.py]: I edited like your suggestion

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 5 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I edited. thank you.

File: purchase_report_custom_hls/__manifest__.py (hunk 3 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/__manifest__.py]: Thank you, I confirmed it works without web depends.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 6 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Thank you, I confirmed work without name.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 5 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I edited it. thank you.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 4 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I delete it. Thank you.

File: purchase_report_custom_hls/__manifest__.py (hunk 3 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/__manifest__.py]: I edited depends, add report_common_hls.

File: purchase_report_custom_hls/models/res_company.py (hunk 7 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/res_company.py]: I edited it.

File: report_common_hls/models/res_company.py (hunk 9 above)
Comment (TomohiroAbe-Quartile) [reply report_common_hls/models/res_company.py]: Sorry, forgot delete. Now deleted it.

File: report_common_hls/views/res_company_views.xml [hunk 10]
```diff
@@ -14,6 +14,14 @@
                     <group string="Common" name="common">
                         <field name="company_chop" widget='image' class="oe_avatar"/>
                     </group>
+                    <group string="Purchase" name="purchase">
```
Comment (yostashiro) [inline report_common_hls/views/res_company_views.xml]: We should not categorize these fields as part of 'purchase' here.  The fields should be moved under 'common'.

File: report_common_hls/i18n/ja_JP.po [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+#. module: report_common_hls
+#: model:ir.model.fields,help:report_common_hls.field_res_company__rep_title
+msgid "The input value is may show as the title of the representative in some of the printed documents."
+msgstr ""
+
+#. module: report_common_hls
+#: model:ir.model.fields,field_description:report_common_hls.field_res_company__rep_title
+msgid "Title of the Representative"
+msgstr "代表名"
```
Comment (yostashiro) [inline report_common_hls/i18n/ja_JP.po]:
```suggestion
msgstr "代表者タイトル"
```

File: report_common_hls/i18n/ja_JP.po [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+#. module: report_common_hls
+#: model:ir.model.fields,field_description:report_common_hls.field_res_company__rep_signature
+msgid "Representative Signature"
+msgstr "代表者署名"
+
+#. module: report_common_hls
+#: model:ir.model.fields,help:report_common_hls.field_res_company__rep_title
+msgid "The input value is may show as the title of the representative in some of the printed documents."
+msgstr ""
```
Comment (yostashiro) [inline report_common_hls/i18n/ja_JP.po]: Translation?

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+                        style="border: none; border-bottom: 2px solid black;"
+                    />
+                    <td width="100%">
+                        <div style="text-align: right; padding-top: 15px padding-right: 10px">
+                            <img
+                                t-if="o.company_id.rep_signature"
+                                t-att-src="image_data_uri(o.company_id.rep_signature)"
+                                style="max-height: 40px;"
+                                alt="Signature"
+                            />
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Why the signature below the border line? It should be above.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+                <strong>Payment Terms:</strong>
+                <br />
+                <span t-field="o.payment_term_id"/>
+            </div>
+        </xpath>
+        <xpath expr="//div[@class='page']" position="after">
+            <div t-if="o.company_id.rep_title">
+                <tr>
+                    <hr
+                        width="100%"
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I suppose this is not needed.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+                        style="border: none; border-bottom: 2px solid black;"
+                    />
+                    <td width="100%">
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I suppose the width attribute is not needed.

Wasn't sure if the `td` tag was necessary (I just don't know).

File: purchase_report_custom_hls/readme/DESCRIPTION.rst [hunk 16]
```diff
@@ -0,0 +1,4 @@
+This module does the following:
+
+- Add some fields in res_company view.
+- The input value will show in the printed purchase order, together with the company information.
```
Comment (yostashiro) [inline purchase_report_custom_hls/readme/DESCRIPTION.rst]:
```suggestion
- Shows the payment terms and the signature image of the representative on the purchase order.

The module depends on the report_common_hls module.
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 13 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Sorry, I edited it.
Layout wasn't same as expected, so added parent class and edit styles.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 14 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Confirmed work as expected.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 15 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I edited div class.
If don't define width with block element, `hr` width will be same as 100vw.

Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I deleted it.
 Thank you.

File: purchase_report_custom_hls/readme/DESCRIPTION.rst (hunk 16 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/readme/DESCRIPTION.rst]: I edited it. Thank you.

File: report_common_hls/i18n/ja_JP.po (hunk 12 above)
Comment (TomohiroAbe-Quartile) [reply report_common_hls/i18n/ja_JP.po]: Sorry, I added.

File: report_common_hls/views/res_company_views.xml (hunk 10 above)
Comment (TomohiroAbe-Quartile) [reply report_common_hls/views/res_company_views.xml]: I move into common group.

File: report_common_hls/i18n/ja_JP.po (hunk 11 above)
Comment (TomohiroAbe-Quartile) [reply report_common_hls/i18n/ja_JP.po]: I edited it.

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
delete tr-tags and add page-break-inside: avoid; in parent div-tag.

**Review by yostashiro — APPROVED** [review]
LGTM 👍

**Pull request merged**